libvips/iofuncs/threadpool.c

```diff
@@ -999,6 +870,12 @@ vips__threadpool_init( void )
 
 	if( g_getenv( "VIPS_STALL" ) )
 		vips__stall = TRUE;
+
+	if( vips__concurrency == 0 )
+		vips__concurrency = vips__concurrency_get_default();
+
+	vips__pool = g_thread_pool_new( vips_thread_main_loop, NULL,
+		-1, FALSE, NULL );
```

Comment:
Should this `-1` be `vips__concurrency`?

Reply:
Ooop, no -1 is correct.

Reply:
AFAIK, such limiting is only necessary for exclusive thread pools, which is [required on WASM](https://github.com/kleisauke/wasm-vips/blob/fc889f48c911ef7e728f0ddfcebdce80be72e70e/build/patches/vips-1492-emscripten.patch), for example.

I also tried this in the past for non-exclusive thread pools (see commits https://github.com/libvips/libvips/commit/c61c59991d890cdeef18f302fee7cab82e3093bc, https://github.com/libvips/libvips/commit/2683368978e907feae675f5147cb80b9412cc1a1 and https://github.com/libvips/libvips/commit/cdbc891205c4e15dcfb04bda74cd19569bf65d22) but I eventually reverted that with commit https://github.com/libvips/libvips/commit/9a9e52cbccd283c59731bbbcefdb458ce2603a2a as it was difficult to do correctly and it didn't work with `dzsave`. The discussion has some more details: https://github.com/libvips/libvips/discussions/2038#discussioncomment-376998.

Reply:
Indeed, the previous `vips_g_thread_join` calls were replaced with a semaphore. Given that we are using the threadpool semantics of GLib now, do you think we should introduce wrappers (or environment variables) for the [`g_thread_pool_set_max_unused_threads`](https://developer.gnome.org/glib/stable/glib-Thread-Pools.html#g-thread-pool-set-max-unused-threads) (default 2) and [`g_thread_pool_set_max_idle_time`](https://developer.gnome.org/glib/stable/glib-Thread-Pools.html#g-thread-pool-set-max-idle-time) (default 15s) functions? This allows users to set the number of unused threads kept alive and the maximum idle time of a thread.

The `vips_threadpool_push` function was named this way because it's just a small wrapper around [`g_thread_pool_push`](https://developer.gnome.org/glib/stable/glib-Thread-Pools.html#g-thread-pool-push). I'm not sure if `vips_thread_new` is better, since it's not directly bound to a thread (as threads can be reused). For the same reason, I renamed `VipsThread` to `VipsTask` within `threadpool.c`. How about renaming it to `vips_task_push` instead? Perhaps we should make this function internal, because I'm not sure if it's intended to be used outside of libvips.

Reply:
OK, let's make it internal, it'll stop me worrying about the name too, heh.

Reply:
... I couldn't resist renaming it to `vips__threadpool_execute()`.

Reply:
Great, thanks!